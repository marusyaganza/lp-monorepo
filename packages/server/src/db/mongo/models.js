const { formatIds, formatFilter } = require('../../utils/dbUtils');
const { ObjectId } = require('mongodb');

function createModel(db, table) {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result = await db.collection(table).find(formatFilter(filter));
    const data = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields) {
    const data = { ...fields };
    delete data.id;
    delete data.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    return result?.result;
  }

  async function deleteOne(filter) {
    console.log('filter', filter);
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    return result?.result;
  }

  async function createMany(entries) {
    const formattedEntries = entries.map(entry => ({
      ...entry,
      createdAt: Date.now()
    }));
    const result = await db.collection(table).insertMany(formattedEntries);
    return result.ops.map(formatIds);
  }

  return {
    findOne,
    findMany,
    createOne,
    createMany,
    updateOne,
    deleteOne
  };
}

module.exports = createModel;
