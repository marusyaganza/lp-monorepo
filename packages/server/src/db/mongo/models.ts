import { formatIds, formatFilter } from '../../utils/dbUtils';

import { Table, ModelType, DbData } from '..';

const { ObjectId } = require('mongodb');

function createModel<T extends DbData>(db: any, table: Table): ModelType<T> {
  async function findOne(filter = {}) {
    const collection = await db.collection(table);
    const result = await collection.findOne(formatFilter(filter));

    return formatIds(result);
  }

  async function findMany(filter = {}) {
    const result: T[] = await db.collection(table).find(formatFilter(filter));
    const data: T[] = [];
    await result.forEach(doc => {
      data.push(formatIds(doc));
    });
    return data;
  }

  async function createOne(fields: Partial<T>) {
    const data = { ...fields, createdAt: Date.now() };
    if (fields?.user) {
      data.user = ObjectId(fields.user);
    }
    const result = await db.collection(table).insertOne(data);

    return formatIds(result.ops[0]);
  }

  async function updateOne(fields: Partial<T>) {
    const data = { ...fields };
    delete data?.id;
    delete data?.user;
    const result = await db
      .collection(table)
      .updateOne(
        { _id: ObjectId(fields.id), user: ObjectId(fields.user) },
        { $set: { ...data } }
      );
    const isOk = result?.result?.ok ? true : false;
    const modifiedCount = result?.result?.nModified;
    return { ok: isOk, modifiedCount };
  }

  async function deleteOne(filter: Partial<T>) {
    const result = await db.collection(table).deleteOne(formatFilter(filter));
    const isOk = result?.result?.ok ? true : false;
    return { ok: isOk };
  }

  return {
    findOne,
    findMany,
    createOne,
    updateOne,
    deleteOne
  };
}

export default createModel;
