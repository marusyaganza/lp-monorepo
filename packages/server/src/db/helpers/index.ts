export * from './sortingHelpers/sortingHelpers';
export * from './general/general';
