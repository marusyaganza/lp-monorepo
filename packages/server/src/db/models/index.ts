import { UserModel, UserModelType } from './User';
import { WordModel, WordModelType } from './Word';
import { WordTagModel, WordTagModelType } from './WordTag';

export interface ModelsType {
  User: UserModelType;
  Word: WordModelType;
  WordTag: WordTagModelType;
}

export const models: ModelsType = {
  User: UserModel,
  Word: WordModel,
  WordTag: WordTagModel
};
