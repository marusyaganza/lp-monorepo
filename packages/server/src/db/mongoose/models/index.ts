import { UserModel } from './User';
import { WordModel } from './Word';

export const models = {
  User: UserModel,
  Word: WordModel
};
