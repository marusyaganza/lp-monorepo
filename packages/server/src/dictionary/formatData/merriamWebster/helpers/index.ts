export * from './general/general';
export * from './resource/resource';
export * from './string/string';
export * from './def/def';
