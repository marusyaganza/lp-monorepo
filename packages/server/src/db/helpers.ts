import { Document, Types } from 'mongoose';
import {
  Maybe,
  Language,
  Game,
  SortBy,
  SortWordsBy
} from '../generated/graphql';

type ModelDataType = {
  id?: Maybe<string>;
  user?: string;
};

interface DbData extends ModelDataType {
  _id?: string;
}

type DocumentType<T> =
  | (Document<unknown, unknown, T> &
      T & {
        _id: Types.ObjectId;
      })
  | null;

export interface WordsFilterType {
  sortBy?: SortBy | SortWordsBy | 'updatedAt';
  language: Language;
  isReverseOrder?: boolean;
  timesToLearn?: number | null;
  gameType?: Game;
  user?: string;
  tags?: string[];
  searchQuery?: string;
}

export interface WordsWithPaginationFilter extends WordsFilterType {
  page?: number;
  limit?: number;
}

export function formatData<T>(data: DocumentType<T>): T | null {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter?.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
const NO_TAGS_ID = '000000000000000000000000';

export function getWordsFilters(filter: WordsFilterType) {
  const {
    user,
    language = Language.English,
    sortBy,
    isReverseOrder,
    gameType,
    tags,
    searchQuery,
    timesToLearn = 5
  } = filter;

  // filter for a specific game
  let gameFilter;
  if (gameType === Game.Conjugation) {
    gameFilter = { conjugation: { $ne: null } };
  }
  if (gameType === Game.Audio) {
    gameFilter = { audioUrl: { $ne: '' } };
  }

  // filter based on tags
  let tagsFilters: { tags: string }[] = [];
  if (tags?.length) {
    tagsFilters = tags
      ?.filter(tag => tag !== NO_TAGS_ID)
      .map(tag => {
        return { tags: tag };
      });
  }

  let noTagsFilter;
  if (tags?.includes(NO_TAGS_ID)) {
    noTagsFilter = { $or: [{ tags: [] }, { tags: null }] };
  }

  let searchFilter;
  if (searchQuery) {
    const searchRegexp = { $regex: new RegExp(searchQuery, 'i') };
    searchFilter = {
      $or: [
        { name: searchRegexp },
        { 'defs.def': searchRegexp },
        { alternativeSpelling: searchRegexp },
        { stems: searchRegexp }
      ]
    };
  }

  // filter based on learning status
  let learningStatusFilters: Record<string, unknown>[] = [];
  if (gameType && sortBy !== SortBy.MemoryRefresher) {
    learningStatusFilters = [
      { isLearned: { $ne: true } },
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $lt: timesToLearn } },
          { [`statistics.${gameType}.successRate`]: null }
        ]
      }
    ];
  }
  if (sortBy === SortBy.MemoryRefresher) {
    learningStatusFilters = [
      {
        $or: [
          { [`statistics.${gameType}.successRate`]: { $gte: timesToLearn } },
          { isLearned: true }
        ]
      }
    ];
  }

  //sorting
  let orderNum = isReverseOrder ? -1 : 1;

  // words with the highest number of errors should go first
  if (sortBy === SortBy.ErrorCount) {
    orderNum = -1 * orderNum;
  }

  let sort: Record<string, number> = { $natural: orderNum };

  if (sortBy) {
    let propName: string = sortBy;
    if (gameType) {
      propName = `statistics.${gameType}.${sortBy}`;
    }

    if (sortBy !== SortBy.MemoryRefresher) {
      sort = { [propName]: orderNum };
    }
  }

  const filtersArray = [
    gameFilter,
    noTagsFilter,
    searchFilter,
    ...tagsFilters,
    ...learningStatusFilters
  ].filter(Boolean);

  const filters: Record<string, unknown> = {
    user,
    language
  };

  if (filtersArray?.length) {
    filters.$and = filtersArray;
  }

  return {
    sort,
    filters
  };
}
