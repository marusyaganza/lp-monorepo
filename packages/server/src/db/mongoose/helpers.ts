import { Types, Document } from 'mongoose';
import { User, Word } from '../../generated/graphql';
import { DbData } from 'db';

type DocumentType =
  | (Document<unknown, Record<string, any>, User | Word> &
      Omit<
        | User
        | (Word & {
            _id: Types.ObjectId;
          }),
        never
      >)
  | null;

export function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({
    getters: true,
    versionKey: false
  });
}

export function formatFilter<T extends DbData>(filter: T): T {
  const id = filter.id;
  if (id) {
    const formattedFilter = { ...filter, _id: id };
    delete formattedFilter.id;
    return formattedFilter;
  }
  return filter;
}
