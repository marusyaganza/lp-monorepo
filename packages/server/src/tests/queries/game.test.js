import { createTestServer, getErrorMessageFromGQL } from '../helpers';
import { Language, Game, SortBy } from '../../generated/graphql';
import { generateGameData } from '../../utils/generateGameData';
import { randomNumbersArray } from '../mocks/gameGenerationData';

import { models } from '../mocks/models';
import { gameQueries } from './queries';

const user = { id: '1' };

let index = 0;

jest.spyOn(global.Math, 'random').mockImplementation(function () {
  if (index === randomNumbersArray.length) {
    index = 0;
  }
  const randomNum = randomNumbersArray[index];
  index++;
  return randomNum;
});

describe('queries', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  test('game queries', async () => {
    const { query } = createTestServer({ models });
    const res = await query({ query: gameQueries.gamesQuery });
    expect(res).toMatchSnapshot();
  });

  test('get gamedata with walid user for Audio game and English', async () => {
    const { query } = createTestServer({
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input: { gameType: Game.Audio, language: Language.English }
      }
    });
    expect(res).toMatchSnapshot();
  });

  test('get gamedata with walid user for SelectDef game and English', async () => {
    const input = {
      gameType: Game.SelectDef,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.PracticedTimes
    };
    const { query } = createTestServer({
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(models.Word.findManyAndSort).toHaveBeenCalledWith({
      ...input,
      timesToLearn: 5,
      user: user.id
    });
    expect(models.Word.findMany).toHaveBeenCalledWith(
      { language: Language.English, user: '1' },
      'name shortDef',
      {
        limit: 40
      }
    );
  });

  test('get gamedata with walid user for SelectWord game and English', async () => {
    const input = {
      gameType: Game.SelectWord,
      language: Language.English,
      isReverseOrder: false,
      sortBy: SortBy.LastTimePracticed
    };

    const { query } = createTestServer({
      generateGameData,
      models,
      user
    });

    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(models.Word.findManyAndSort).toHaveBeenCalledWith({
      ...input,
      timesToLearn: 5,
      user: user.id
    });
    expect(models.Word.findMany).toHaveBeenCalledWith(
      { language: Language.English, user: '1' },
      'name shortDef',
      {
        limit: 40
      }
    );
  });

  test('get gamedata with walid user for TypeWord game and English', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(models.Word.findManyAndSort).toHaveBeenCalledWith({
      ...input,
      timesToLearn: 5,
      user: user.id
    });
  });

  test('get gamedata with walid user for Conjugation game and Spanish', async () => {
    const input = {
      gameType: Game.Conjugation,
      language: Language.Spanish,
      sortBy: SortBy.SuccessRate
    };
    const { query } = createTestServer({
      generateGameData,
      models,
      user
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(res).toMatchSnapshot();
    expect(models.Word.findManyAndSort).toHaveBeenCalledWith({
      ...input,
      timesToLearn: 5,
      user: user.id
    });
  });

  test('should not generate game data if user is not found', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      models: {
        ...models,
        User: {
          findOne: jest.fn()
        }
      },
      user
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'Login to perform this operation'
    );
    expect(models.Word.findManyAndSort).toHaveBeenCalledTimes(0);
  });

  test('should not generate game data if user is undefined', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      models
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'Login to perform this operation'
    );
    expect(models.Word.findManyAndSort).toHaveBeenCalledTimes(0);
  });

  test('generate game data if user does not have enough words', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn(() => [])
        }
      }
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'not enough words to start a game. You have 0 word. Words requited for the game: 1'
    );
  });

  test('generate game data if words do not exist', async () => {
    const input = {
      gameType: Game.TypeWord,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn()
        }
      }
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual('words not found');
  });

  test('generate game data if game is not found', async () => {
    const input = {
      gameType: Game.SelectDef,
      language: Language.English,
      isReverseOrder: true,
      sortBy: SortBy.ErrorCount
    };
    const { query } = createTestServer({
      generateGameData,
      user,
      models: {
        ...models,
        Word: {
          findManyAndSort: jest.fn()
        },
        Game: {
          findOne: jest.fn()
        }
      }
    });
    const res = await query({
      query: gameQueries.gameQuery,
      variables: {
        input
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual('game not found');
  });

  test('generate game data if game is not available for a language', async () => {
    const { query } = createTestServer({
      generateGameData,
      user,
      models: {
        ...models,
        Game: {
          ...models.Game,
          findMany: jest.fn()
        }
      }
    });
    const res = await query({
      query: gameQueries.gamesQuery,
      variables: {
        language: Language.English
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'no games for ENGLISH language were found'
    );
  });
});
