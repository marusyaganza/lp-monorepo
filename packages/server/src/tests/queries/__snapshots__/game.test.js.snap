// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries game queries 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "games": [
          {
            "desc": "Listen, Type, Master. Sharpen language skills through auditory engagement.",
            "id": "audio",
            "name": "WordEcho",
          },
          {
            "desc": "Define & Conquer.Match words with their meanings intuitively.",
            "id": "select_definition",
            "name": "LexiSense",
          },
          {
            "desc": "Associate definitions with the right words. A captivating word puzzle adventure!",
            "id": "select_word",
            "name": "Definify: Match & Learn",
          },
          {
            "desc": "Type words matching given definitions. Boost vocabulary in engaging rounds!",
            "id": "type_word",
            "name": "WordRecall: Define & Type",
          },
          {
            "desc": "Conjugate verbs!",
            "id": "conjugation",
            "name": "Conjugation",
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries get gamedata with walid user for Audio game and English 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "game": {
          "questions": [
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3",
              ],
              "wordId": "1",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v1.mp3",
              ],
              "wordId": "2",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel and axle",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v3.mp3",
              ],
              "wordId": "3",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel animal",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v4.mp3",
              ],
              "wordId": "4",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel bug",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v5.mp3",
              ],
              "wordId": "5",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel-thrown",
              "options": null,
              "question": [
                "https://media.merriam-webster.com/audio/prons/en/us/mp3/number/9wheel01.mp3",
              ],
              "wordId": "7",
            },
          ],
          "task": "Type the word that you've heard",
          "type": "AUDIO",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries get gamedata with walid user for Conjugation game and Spanish 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "game": {
          "questions": [
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "caigo, caes, cae, caemos, caÃ©is, caen",
              "options": null,
              "question": [
                "caer",
              ],
              "wordId": "0",
            },
          ],
          "task": "Conjugate the verb in Presente indicativo",
          "type": "CONJUGATION",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries get gamedata with walid user for SelectDef game and English 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "game": {
          "questions": [
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3",
              },
              "answer": "a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle",
              ],
              "question": [
                "wheel",
              ],
              "wordId": "1",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v1.mp3",
              },
              "answer": "to turn on or as if on an axis : revolve",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "to turn on or as if on an axis : revolve",
              ],
              "question": [
                "wheel",
              ],
              "wordId": "2",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v3.mp3",
              },
              "answer": "a mechanical device consisting of a grooved wheel turned by a cord or chain with a rigidly attached axle (as for winding up a weight) together with the supporting standards",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "a mechanical device consisting of a grooved wheel turned by a cord or chain with a rigidly attached axle (as for winding up a weight) together with the supporting standards",
              ],
              "question": [
                "wheel and axle",
              ],
              "wordId": "3",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v4.mp3",
              },
              "answer": "rotifer",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "rotifer",
              ],
              "question": [
                "wheel animal",
              ],
              "wordId": "4",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v5.mp3",
              },
              "answer": "a large North American assassin bug (Arilus cristatus) that has a high serrated crest on its prothorax and preys on insects",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "a large North American assassin bug (Arilus cristatus) that has a high serrated crest on its prothorax and preys on insects",
              ],
              "question": [
                "wheel bug",
              ],
              "wordId": "5",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "a gun's lock for a muzzle-loading firearm in which sparks are struck from a flint or a piece of pyrite by a revolving wheel",
              "options": [
                "",
                "delightful feeling of anticipation and excitement that arises before embarking on a new adventure or journey.",
                "to engage in a playful and creative activity that sparks joy and imagination. It involves letting go of inhibitions and embracing a carefree spirit to explore new ideas and possibilities",
                "a whimsical, fleeting moment when sunlight catches water droplets in the air, creating a shimmering and enchanting effect",
                "a gentle, soothing sensation experienced when a cool breeze rustles through leaves on a warm summer day",
                "a gun's lock for a muzzle-loading firearm in which sparks are struck from a flint or a piece of pyrite by a revolving wheel",
              ],
              "question": [
                "wheel lock",
              ],
              "wordId": "6",
            },
          ],
          "task": "Select a definition that means ",
          "type": "SELECT_DEF",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries get gamedata with walid user for SelectWord game and English 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "game": {
          "questions": [
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3",
              },
              "answer": "wheel",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel",
              ],
              "question": [
                "a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle",
                "a contrivance or apparatus having as its principal part a wheel: such as",
                "a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)",
              ],
              "wordId": "1",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v1.mp3",
              },
              "answer": "wheel",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel",
              ],
              "question": [
                "to turn on or as if on an axis : revolve",
                "to change direction as if revolving on a pivot",
                "to move or extend in a circle or curve",
              ],
              "wordId": "2",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v3.mp3",
              },
              "answer": "wheel and axle",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel and axle",
              ],
              "question": [
                "a mechanical device consisting of a grooved wheel turned by a cord or chain with a rigidly attached axle (as for winding up a weight) together with the supporting standards",
              ],
              "wordId": "3",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v4.mp3",
              },
              "answer": "wheel animal",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel animal",
              ],
              "question": [
                "rotifer",
              ],
              "wordId": "4",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v5.mp3",
              },
              "answer": "wheel bug",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel bug",
              ],
              "question": [
                "a large North American assassin bug (Arilus cristatus) that has a high serrated crest on its prothorax and preys on insects",
              ],
              "wordId": "5",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel lock",
              "options": [
                "egalitarian",
                "quix",
                "flibberfazzle",
                "glitzephyr",
                "zephral",
                "wheel lock",
              ],
              "question": [
                "a gun's lock for a muzzle-loading firearm in which sparks are struck from a flint or a piece of pyrite by a revolving wheel",
              ],
              "wordId": "6",
            },
          ],
          "task": "Select a word that means ",
          "type": "SELECT_WORD",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`queries get gamedata with walid user for TypeWord game and English 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "game": {
          "questions": [
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3",
              },
              "answer": "wheel",
              "options": null,
              "question": [
                "a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle",
                "a contrivance or apparatus having as its principal part a wheel: such as",
                "a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)",
              ],
              "wordId": "1",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v1.mp3",
              },
              "answer": "wheel",
              "options": null,
              "question": [
                "to turn on or as if on an axis : revolve",
                "to change direction as if revolving on a pivot",
                "to move or extend in a circle or curve",
              ],
              "wordId": "2",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v3.mp3",
              },
              "answer": "wheel and axle",
              "options": null,
              "question": [
                "a mechanical device consisting of a grooved wheel turned by a cord or chain with a rigidly attached axle (as for winding up a weight) together with the supporting standards",
              ],
              "wordId": "3",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v4.mp3",
              },
              "answer": "wheel animal",
              "options": null,
              "question": [
                "rotifer",
              ],
              "wordId": "4",
            },
            {
              "additionalInfo": {
                "audioUrl": "https://media.merriam-webster.com/audio/prons/en/us/mp3/a/all_w01v5.mp3",
              },
              "answer": "wheel bug",
              "options": null,
              "question": [
                "a large North American assassin bug (Arilus cristatus) that has a high serrated crest on its prothorax and preys on insects",
              ],
              "wordId": "5",
            },
            {
              "additionalInfo": {
                "audioUrl": null,
              },
              "answer": "wheel lock",
              "options": null,
              "question": [
                "a gun's lock for a muzzle-loading firearm in which sparks are struck from a flint or a piece of pyrite by a revolving wheel",
              ],
              "wordId": "6",
            },
          ],
          "task": "Type a word that means ",
          "type": "TYPE_WORD",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
